File: backend/app/models/publisher.py
from typing import Any, Dict, List, Optional
from uuid import UUID

from app.config.database import database


class Publisher:
    @staticmethod
    async def create(publisher_data: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        query = """
            INSERT INTO publishers (name, name_th, description, description_th, website_url, logo_url)
            VALUES (:name, :name_th, :description, :description_th, :website_url, :logo_url)
            RETURNING *
        """
        values = {
            "name": publisher_data.get("name"),
            "name_th": publisher_data.get("name_th"),
            "description": publisher_data.get("description"),
            "description_th": publisher_data.get("description_th"),
            "website_url": publisher_data.get("website_url"),
            "logo_url": publisher_data.get("logo_url"),
        }
        result = await database.fetch_one(query=query, values=values)
        return dict(result) if result else None

    @staticmethod
    async def find_by_id(publisher_id: UUID) -> Optional[Dict[str, Any]]:
        query = """
            SELECT p.*, 
                   (SELECT COUNT(*) FROM books WHERE publisher_id = p.id) as book_count
            FROM publishers p
            WHERE p.id = :id
        """
        result = await database.fetch_one(query=query, values={"id": str(publisher_id)})
        return dict(result) if result else None

    @staticmethod
    async def update(publisher_id: UUID, publisher_data: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        query = """
            UPDATE publishers 
            SET name = COALESCE(:name, name),
                name_th = COALESCE(:name_th, name_th),
                description = COALESCE(:description, description),
                description_th = COALESCE(:description_th, description_th),
                website_url = COALESCE(:website_url, website_url),
                logo_url = COALESCE(:logo_url, logo_url)
            WHERE id = :id
            RETURNING *
        """
        values = {
            "id": str(publisher_id),
            "name": publisher_data.get("name"),
            "name_th": publisher_data.get("name_th"),
            "description": publisher_data.get("description"),
            "description_th": publisher_data.get("description_th"),
            "website_url": publisher_data.get("website_url"),
            "logo_url": publisher_data.get("logo_url"),
        }
        result = await database.fetch_one(query=query, values=values)
        return dict(result) if result else None

    @staticmethod
    async def delete(publisher_id: UUID) -> Optional[Dict[str, Any]]:
        query = "DELETE FROM publishers WHERE id = :id RETURNING *"
        result = await database.fetch_one(query=query, values={"id": str(publisher_id)})
        return dict(result) if result else None

    @staticmethod
    async def get_all(page: int = 1, limit: int = 20) -> Dict[str, Any]:
        offset = (page - 1) * limit
        query = """
            SELECT p.*, 
                   (SELECT COUNT(*) FROM books WHERE publisher_id = p.id) as book_count
            FROM publishers p
            ORDER BY p.name ASC
            LIMIT :limit OFFSET :offset
        """
        count_query = "SELECT COUNT(*) as count FROM publishers"
        
        results = await database.fetch_all(query=query, values={"limit": limit, "offset": offset})
        count_result = await database.fetch_one(query=count_query)
        
        total = count_result["count"] if count_result else 0
        return {
            "publishers": [dict(r) for r in results],
            "total": total,
            "page": page,
            "totalPages": (total + limit - 1) // limit if total > 0 else 0,
        }

    @staticmethod
    async def search(search_query: str, limit: int = 10) -> List[Dict[str, Any]]:
        query = """
            SELECT p.*, 
                   (SELECT COUNT(*) FROM books WHERE publisher_id = p.id) as book_count
            FROM publishers p
            WHERE p.name ILIKE :search_query OR p.name_th ILIKE :search_query
            ORDER BY p.name ASC
            LIMIT :limit
        """
        results = await database.fetch_all(
            query=query, 
            values={"search_query": f"%{search_query}%", "limit": limit}
        )
        return [dict(r) for r in results]

    @staticmethod
    async def get_books(publisher_id: UUID, page: int = 1, limit: int = 20) -> List[Dict[str, Any]]:
        offset = (page - 1) * limit
        query = """
            SELECT b.*
            FROM books b
            WHERE b.publisher_id = :publisher_id
            ORDER BY b.created_at DESC
            LIMIT :limit OFFSET :offset
        """
        results = await database.fetch_all(
            query=query, 
            values={"publisher_id": str(publisher_id), "limit": limit, "offset": offset}
        )
        return [dict(r) for r in results]

File: backend/src/models/Publisher.js
const db = require('../config/database');

class Publisher {
  static async create(publisherData) {
    const { name, name_th, description, description_th, website_url, logo_url } = publisherData;
    const query = `
      INSERT INTO publishers (name, name_th, description, description_th, website_url, logo_url)
      VALUES ($1, $2, $3, $4, $5, $6)
      RETURNING *
    `;
    const values = [name, name_th, description, description_th, website_url, logo_url];
    const result = await db.query(query, values);
    return result.rows[0];
  }

  static async findById(id) {
    const query = `
      SELECT p.*, 
             (SELECT COUNT(*) FROM books WHERE publisher_id = p.id) as book_count
      FROM publishers p
      WHERE p.id = $1
    `;
    const result = await db.query(query, [id]);
    return result.rows[0];
  }

  static async update(id, publisherData) {
    const { name, name_th, description, description_th, website_url, logo_url } = publisherData;
    const query = `
      UPDATE publishers 
      SET name = COALESCE($1, name),
          name_th = COALESCE($2, name_th),
          description = COALESCE($3, description),
          description_th = COALESCE($4, description_th),
          website_url = COALESCE($5, website_url),
          logo_url = COALESCE($6, logo_url)
      WHERE id = $7
      RETURNING *
    `;
    const values = [name, name_th, description, description_th, website_url, logo_url, id];
    const result = await db.query(query, values);
    return result.rows[0];
  }

  static async delete(id) {
    const query = 'DELETE FROM publishers WHERE id = $1 RETURNING *';
    const result = await db.query(query, [id]);
    return result.rows[0];
  }

  static async getAll(page = 1, limit = 20) {
    const offset = (page - 1) * limit;
    const query = `
      SELECT p.*, 
             (SELECT COUNT(*) FROM books WHERE publisher_id = p.id) as book_count
      FROM publishers p
      ORDER BY p.name ASC
      LIMIT $1 OFFSET $2
    `;
    const countQuery = 'SELECT COUNT(*) FROM publishers';
    const [result, countResult] = await Promise.all([
      db.query(query, [limit, offset]),
      db.query(countQuery)
    ]);
    return {
      publishers: result.rows,
      total: parseInt(countResult.rows[0].count),
      page,
      totalPages: Math.ceil(countResult.rows[0].count / limit)
    };
  }

  static async search(searchQuery, limit = 10) {
    const query = `
      SELECT p.*, 
             (SELECT COUNT(*) FROM books WHERE publisher_id = p.id) as book_count
      FROM publishers p
      WHERE p.name ILIKE $1 OR p.name_th ILIKE $1
      ORDER BY p.name ASC
      LIMIT $2
    `;
    const result = await db.query(query, [`%${searchQuery}%`, limit]);
    return result.rows;
  }

  static async getBooks(publisherId, page = 1, limit = 20) {
    const offset = (page - 1) * limit;
    const query = `
      SELECT b.*
      FROM books b
      WHERE b.publisher_id = $1
      ORDER BY b.created_at DESC
      LIMIT $2 OFFSET $3
    `;
    const result = await db.query(query, [publisherId, limit, offset]);
    return result.rows;
  }
}

module.exports = Publisher;
